File: scripts/verify_dwug_final.py
"""Export exact saved DWUG predictions for independent base-R rank verification."""
import argparse
import hashlib
import json
from pathlib import Path
import subprocess
import numpy as np
import pandas as pd


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--analysis', type=Path, required=True)
    parser.add_argument('--output', type=Path, required=True)
    args = parser.parse_args()
    args.output.mkdir(parents=True, exist_ok=False)
    gold = pd.read_csv(args.analysis/'aggregated-pair-gold.csv')
    predictions = np.load(args.analysis/'pair-predictions.npz')
    expected = json.loads((args.analysis/'pooled-pair-validity.json').read_text())
    rows = []
    for method in predictions.files:
        values = predictions[method]
        if values.shape != (len(gold),) or not np.isfinite(values).all():
            raise ValueError('Invalid saved pair coverage.')
        values.astype('<f8').tofile(args.output/(method+'.bin'))
        rows.append({'method': method, 'spearman': expected[method]['spearman']})
    gold.judgment.to_numpy(dtype='<f8').tofile(args.output/'gold.bin')
    pd.DataFrame(rows).to_csv(args.output/'expected.csv', index=False)
    script = Path(__file__).with_suffix('.R')
    subprocess.run(['Rscript', str(script), str(args.analysis), str(args.output), str(len(gold))], check=True)
    sources = [args.analysis/name for name in ['aggregated-pair-gold.csv', 'pair-predictions.npz',
        'pooled-pair-validity.json', 'per-target-shift.csv', 'change-validity.csv']]
    sources += [Path(__file__), script]
    (args.output/'manifest.json').write_text(json.dumps({'status': 'passed',
        'scope': 'Independent R recomputation of 8 pooled pair and 32 across-target rank correlations; does not independently validate input feature extraction or confidence intervals',
        'n_pairs': len(gold), 'sha256': {str(p): hashlib.sha256(p.read_bytes()).hexdigest() for p in sources}}, indent=2))


if __name__ == '__main__':
    main()

File: scripts/verify_dwug_final.R
args <- commandArgs(trailingOnly=TRUE)
analysis <- args[1]; output <- args[2]; n <- as.integer(args[3])
expected <- read.csv(file.path(output, 'expected.csv'), check.names=FALSE)
gold <- readBin(file.path(output, 'gold.bin'), 'double', n=n, size=8, endian='little')
rows <- list()
for (i in seq_len(nrow(expected))) {
  method <- expected$method[i]
  pred <- readBin(file.path(output, paste0(method, '.bin')), 'double', n=n, size=8, endian='little')
  stopifnot(length(pred) == n, length(gold) == n)
  actual <- cor(gold, pred, method='spearman')
  rows[[length(rows)+1]] <- data.frame(method=method, metric='pooled_pair', expected=expected$spearman[i], actual=actual)
}
shift <- read.csv(file.path(analysis, 'per-target-shift.csv'), check.names=FALSE)
change <- read.csv(file.path(analysis, 'change-validity.csv'), check.names=FALSE)
for (i in seq_len(nrow(change))) {
  row <- change[i, ]; subset <- shift[shift$method == row$method, ]
  stopifnot(nrow(subset) == row$n_targets, !anyDuplicated(subset$target))
  actual <- cor(subset$human_change, subset[[row$metric]], method='spearman')
  rows[[length(rows)+1]] <- data.frame(method=row$method, metric=row$metric, expected=row$spearman, actual=actual)
}
result <- do.call(rbind, rows)
result$error <- abs(result$actual-result$expected)
write.csv(result, file.path(output, 'independent-correlations.csv'), row.names=FALSE)
stopifnot(all(is.finite(result$error)), max(result$error) < 1e-12)
cat(nrow(result), 'correlations verified; max error', max(result$error), '\n')
